### ISSUE TITLE
ActionController::UnknownHttpMethod does not get automatically rescued as documented

### ISSUE THREAD

**Issue by lorennorman** [rails/rails#38998]
**TL;DR:** `Rails::Rack::Logger` will raise `ActionController::UnknownHttpMethod` and appears earlier in the middleware stack than `ActionDispatch:ShowExceptions` or `DebugExceptions`, meaning we get 500s and our exception notifiers fire, instead of getting the quiet, documented 405s.

### Steps to reproduce
Here's an integration test snippet that demonstrates the issue (make sure to call a route that exists):
```ruby
class UnknownHttpMethodTest < ActionDispatch::IntegrationTest
  test 'invalid http method should 405' do
    process "NOT_A_METHOD", '/'
    assert_response :method_not_allowed
  end
end
```

### Expected behavior
According to the Rails Guides, `ActionController::UnknownHttpMethod` should be automatically rescued and respond with `:method_not_allowed`, or HTTP 405.

From [the Guides](https://guides.rubyonrails.org/v6.0/configuring.html#configuring-action-dispatch):
```ruby
config.action_dispatch.rescue_responses = {
  'ActionController::RoutingError' => :not_found,
  'AbstractController::ActionNotFound' => :not_found,
  'ActionController::MethodNotAllowed' => :method_not_allowed,
  'ActionController::UnknownHttpMethod' => :method_not_allowed,
  'ActionController::NotImplemented' => :not_implemented,
  'ActionController::UnknownFormat' => :not_acceptable,
  'ActionController::InvalidAuthenticityToken' => :unprocessable_entity,
  'ActionController::InvalidCrossOriginRequest' => :unprocessable_entity,
  'ActionDispatch::Http::Parameters::ParseError' => :bad_request,
  'ActionController::BadRequest' => :bad_request,
  'ActionController::ParameterMissing' => :bad_request,
  'Rack::QueryParser::ParameterTypeError' => :bad_request,
  'Rack::QueryParser::InvalidParameterError' => :bad_request,
  'ActiveRecord::RecordNotFound' => :not_found,
  'ActiveRecord::StaleObjectError' => :conflict,
  'ActiveRecord::RecordInvalid' => :unprocessable_entity,
  'ActiveRecord::RecordNotSaved' => :unprocessable_entity
}
```

### Actual behavior
The exception is not rescued by Rails and we end up getting a 500 and notifying our team unnecessarily. (Case in point, [this StackOverflow post](https://stackoverflow.com/questions/55639558/how-to-silence-actioncontrollerunknownhttpmethod-errors) talking about their exception notifiers blowing up when bots come through, and getting no quality answers.)

`Rails::Rack::Logger` creates an `ActionDispatch::Request` object to wrap the request, then uses it to produce pretty logs, as seen here:
https://github.com/rails/rails/blob/master/railties/lib/rails/rack/logger.rb#L48-L54
```ruby
        def started_request_message(request) # :doc:
          'Started %s "%s" for %s at %s' % [
            request.request_method,
            request.filtered_path,
            request.remote_ip,
            Time.now.to_default_s ]
        end
```

Unfortunately, `request.request_method` raises the error in question.

Why isn't it rescued? Here's the default middleware stack:
```ruby
use Rack::Sendfile
use ActionDispatch::Static
use ActionDispatch::Executor
use ActiveSupport::Cache::Strategy::LocalCache::Middleware
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use ActionDispatch::RemoteIp
use Sprockets::Rails::QuietAssets
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use WebConsole::Middleware
use ActionDispatch::DebugExceptions
use ActionDispatch::Reloader
use ActionDispatch::Callbacks
use ActiveRecord::Migration::CheckPending
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use ActionDispatch::ContentSecurityPolicy::Middleware
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Rack::TempfileReaper
run MyApp::Application.routes
```

`Rails::Rack::Logger` appears before any of the automatic exception handling middleware (`ActionDispatch::ShowExceptions` or `ActionDispatch::DebugExceptions`)

### Possible fixes

- `Rails::Rack::Logger` could rescue specific exceptions from `ActionDispatch::Request` in order to still log reasonably and let later middleware deal with the problem.
- the middleware stack could be re-ordered to catch this appropriately

I would be happy to work on this and submit a PR with a nudge in the right direction.

### System configuration
**Rails version**: `6.0.2.2`

**Ruby version**: `2.6.3p62`

**Comment by eileencodes** [comment]
Hi @lorennorman thanks for the report. There's not enough information here to determine whether your application is raising something other than `ActionController::UnknownHttpMethod` and that's why you're seeing the 500 or if Rails isn't working correctly.

To help us debug can you provide a script using one of our templates here https://github.com/rails/rails/tree/master/guides/bug_report_templates to reproduce or create a new application that demonstrates the issue? 

Thanks!

**Comment by lorennorman** [comment]
@eileencodes Challenge Accepted! Took a sec to figure out how to actually send a bad HTTP verb, but I feel pretty confident this proves the issue:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", "6.0.2.2"
end

require "rack/test"
require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    render plain: "Home"
  end
end

require "minitest/autorun"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_unknown_http_method_returns_405
    # This raises ActionController::UnknownHttpMethod instead of rescuing it
    # and responding with a 405 as the docs state it should
    request "/", { "REQUEST_METHOD" => "NOT_AN_HTTP_METHOD" }
    # Doesn't get here at all because the raise is not properly rescued due
    # to the ActionDispatch::ShowExceptions middleware coming later in the stack
    # than this raise happens (which is at Rails::Rack::Logger)
    assert_equal 405, last_response.status
  end

  private
    def app
      Rails.application
    end
end
```

Thank you so much for pointing me at these templates, they are sort of amazing.

**Comment by davidmoseler** [comment]
I think the essence of the problem might be the use of request.request_method for printing in the logger. If request.request_method raises an exception when the http request method is not allowed, then it is not the proper function to evoke there. When logging, you only want to find out what was written in the client request, you don't want to raise an ActionController::UnknownHttpMethod just yet. Maybe a raw version of request.request_method should be used instead.

**Comment by lorennorman** [comment]
@davidmoseler excellent information, thank you for sharing.

I agree with your thoughts about changing the order of the middleware vs fixing the logger.

What about adding begin/rescue to the logger?

**Comment by davidmoseler** [comment]
I've created a proposed solution, but, even if everyone agrees that it is the right path to go, there is a lot of work to be done there.

I thought about what you said about an obviously correct thing, and I also don't see it, but I do think an obviously wrong thing is to do checking and raise an exception in an unexpected moment such as when reading "request_method" for the first time.

It seems to me that either the checking should be done in the constructor, or it should be explicitly done latter on. I tried doing it in the constructor and passing an extra option for not doing the checking (as when using the class as a wrapper for logging), but checking the request in the constructor blew up so many dozens of tests that it seems like a very difficult solution.

To do explicit checking only breaks the tests which expect that particular exception to be raised (so the explicit checking should be added to those tests). But it still remains to determine the places where the explicit checking should be done in the entire codebase.

This is my first time contributing (or trying to), so the pull request was just something I ended up doing after following the contributing guidelines for getting used to it, so don't take my solution too seriously and please point me to a better approach if I'm violating some norm.

**Comment by lorennorman** [comment]
Hello, Rails team!

This bug continues to exist, and this issue still explains it well and reproduces it with a failing test.

I've not attached a PR because I wanted some feedback from a core team member about the right approach, it touches some deeper middleware/logging/organization topics that I'm less versed in. I can always "go boldly forward" with something if that's preferred!

**Comment by eugeneius** [comment]
I commented on https://github.com/rails/rails/pull/39041 with a suggested fix, and I'd like to give @davidmoseler a chance to implement it if they're still interested, since they went boldly forward already 🖖

**Comment by eugeneius** [comment]
I think we've waited long enough that you can pick this up @lorennorman, if you're still interested.

**Comment by lorennorman** [comment]
@eugeneius I'm interested!

However, some core team feedback would still help, here's the fixes I suggested:

> ### Possible fixes
>
> - Rails::Rack::Logger could rescue specific exceptions from ActionDispatch::Request in order to still log reasonably and let later middleware deal with the problem.
> - the middleware stack could be re-ordered to catch this appropriately

In the absence of guidance I'll just start trying things and see what breaks the least. 😄

**Comment by eugeneius** [comment]
What do you think of my suggestion in https://github.com/rails/rails/pull/39041#issuecomment-664110607:

> How about aliasing Rack's version of the method before we override it?

It seems like the fix least likely to cause unintended side effects.

**Comment by lorennorman** [comment]
@eugeneius Hey, that worked! Pretty simple, too.

I'll write a test that proves the problem and the fix, anything else you'd like to see in this PR?

**Comment by eugeneius** [comment]
Other than the test, a changelog entry would be good.

**Comment by lorennorman** [comment]
Just pushed a dead-simple test in Railties that makes the problem obvious.

It's difficult because this is a cross-cutting concern: the problem/fix was in ActionPack (request and logger), but the test was only really "in context" in Railties. Is this alright?

Then I wonder which changelog I should be editing? Will do ActionPack, that's where the modified behavior is. Probably wrap this up tomorrow if there are no concerns with all this.

**Comment by eugeneius** [comment]
That test looks good! 👌 

The logger is [defined in Railties](https://github.com/rails/rails/blob/315f8b46038d0ce8ba5ec11e8909715b21be221b/railties/lib/rails/rack/logger.rb), so I think that's the right place for the changelog entry too. Action Pack was doing the right thing; the problem was either in the logger or the default middleware stack, depending on how you look at it.

**Comment by lorennorman** [comment]
Done and done, can't thank you enough for the guidance on every step of this!

**Comment by agrobbin** [comment]
@eugeneius @lorennorman we just ran into this today as well, and I think we might've found another case where this exception is raised at an inopportune time: when this exception itself is raised and `config.exceptions_app = self.routes`:

```
Started TEST "/" for 127.0.0.1 at 2020-11-05 15:40:31 -0500
   (1.0ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
TEST, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH excluded from capture: DSN not set
  
ActionController::UnknownHttpMethod (TEST, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH):
  
actionpack (6.0.3.4) lib/action_dispatch/http/request.rb:431:in `check_method'
actionpack (6.0.3.4) lib/action_dispatch/http/request.rb:143:in `request_method'
rack (2.2.3) lib/rack/request.rb:187:in `head?'
actionpack (6.0.3.4) lib/action_dispatch/journey/router.rb:113:in `find_routes'
actionpack (6.0.3.4) lib/action_dispatch/journey/router.rb:32:in `serve'
actionpack (6.0.3.4) lib/action_dispatch/routing/route_set.rb:834:in `call'
# ...
Error during failsafe response: TEST, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/http/request.rb:431:in `check_method'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/http/request.rb:143:in `request_method'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rack-2.2.3/lib/rack/request.rb:187:in `head?'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/journey/router.rb:113:in `find_routes'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/journey/router.rb:32:in `serve'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/routing/route_set.rb:834:in `call'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/middleware/show_exceptions.rb:50:in `render_exception'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/middleware/show_exceptions.rb:36:in `rescue in call'
  /usr/local/var/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.4/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
  # ...
```

This is due to `exceptions_app` being used in [`ActionDispatch::ShowExceptions#render_exception`](https://github.com/rails/rails/blob/a17c5ff5597f6d77c70d65f82201d8e09b37b445/actionpack/lib/action_dispatch/middleware/show_exceptions.rb#L50), which tries to call various methods on `ActionDispatch::Request`, including (but not limited to) [`head?`](https://github.com/rails/rails/blob/a17c5ff5597f6d77c70d65f82201d8e09b37b445/actionpack/lib/action_dispatch/journey/router.rb#L115-L119), in turn raising another `UnknownHttpMethod` exception.

This could definitely be a can of worms, but the thought I had to try and fix this once and for all was to create a completely separate middleware class, `ActionDispatch::UnknownHttpMethod`, that checks the validity of `request_method` and does the appropriate raising of an exception (inserted after `ActionDispatch::ShowExceptions` in the middleware stack).

`ActionDispatch::Request` would then be able to simply read the `REQUEST_METHOD` header (or the override value from `rack.methodoverride.original_method`) without doing any subsequent checks itself. 

This is still pseudo-code-ish, but something along the lines of this:

```ruby
module ActionDispatch
  class UnknownHttpMethod
    def initialize(app)
      @app = app
    end

    def call(env)
      request = ActionDispatch::Request.new env

      unless valid_request_method?(request.request_method)
        raise(ActionController::UnknownHttpMethod, "#{request.request_method}, accepted HTTP methods are #{ActionDispatch::Request::HTTP_METHODS[0...-1].join(', ')}, and #{ActionDispatch::Request::HTTP_METHODS[-1]}")
      end

      @app.call(env)
    end

    private

    def valid_request_method?(name)
      # ...
    end
  end
end
```

The primary thing I'm not sure about is whether `ActionDispatch::Request` *has* to check validity of `request_method` in other situations (like in the [`request_method=` writer](https://github.com/rails/rails/blob/a17c5ff5597f6d77c70d65f82201d8e09b37b445/actionpack/lib/action_dispatch/http/request.rb#L168-L172) and [`method` reader](https://github.com/rails/rails/blob/a17c5ff5597f6d77c70d65f82201d8e09b37b445/actionpack/lib/action_dispatch/http/request.rb#L198-L203)). @eugeneius do you have guidance on this topic? I am happy to take a pass on implementing this change if you think it's appropriate!

**Comment by newfylox** [comment]
Using Rails 7.0.8 and just got this error message by Rollbar as a 500 error.

```ruby
ActionController::UnknownHttpMethod CNAI, accepted HTTP methods are OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, CONNECT, PROPFIND, PROPPATCH, MKCOL, COPY, MOVE, LOCK, UNLOCK, VERSION-CONTROL, REPORT, CHECKOUT, CHECKIN, UNCHECKOUT, MKWORKSPACE, UPDATE, LABEL, MERGE, BASELINE-CONTROL, MKACTIVITY, ORDERPATCH, ACL, SEARCH, MKCALENDAR, and PATCH
```

Any idea why?
